File: Merge_Sort.py
import random
def merge_sort(nums:list)->list:
    if len(nums) <= 1:
        return nums
    left  = merge_sort(nums[ :len(nums)//2])
    right = merge_sort(nums[len(nums)//2 :])
    point_left  = 0
    point_right = 0
    sorted_nums = []
    while point_left < len(left) and point_right < len(right):
        if left[point_left] < right[point_right]:
            sorted_nums.append(left[point_left])
            point_left += 1
        else:
            sorted_nums.append(right[point_right])
            point_right += 1
    if point_left == len(left):
        sorted_nums += right[point_right:]
    elif point_right == len(right):
        sorted_nums += left[point_left:]
    return sorted_nums

def main():
    unsorted_list = [random.randint(0, 100) for _ in range(10)]
    # unsorted_list = [12, 11, 13, 5, 6, 7]
    print("Before Selection Sort\t: ", unsorted_list)
    print("After Selevtion Sort\t: ", merge_sort(unsorted_list))

if __name__ =="__main__":
    main()

File: mergeSort.cpp
#include<iostream>
#include<vector>

using namespace std;

void merge(vector<int> &nums, int start, int middle, int end){
    vector<int> left(nums.begin()+start, nums.begin()+middle+1);
    vector<int> right(nums.begin()+middle+1, nums.begin()+end+1);
    int indexLeft = 0;
    int indexRight = 0;
    for(int i=start; i<=end; i++){
        if(indexLeft==left.size()){
            nums[i] = right[indexRight];
            indexRight++;
        }else if(indexRight==right.size()){
            nums[i] = left[indexLeft];
            indexLeft++;
        }else if(left[indexLeft]<right[indexRight]){
            nums[i] = left[indexLeft];
            indexLeft++;
        }else{
            nums[i] = right[indexRight];
            indexRight++;
        }
    }
}

void mergeSort(vector<int> &nums, int start, int end){
    if(start==end)
        return;
    int middle = (start+end)/2;
    mergeSort(nums, start, middle);
    mergeSort(nums, middle+1, end);
    merge(nums, start, middle, end);
}

int main(){
    vector<int> nums = {6, 5, 1, 3, 4, 9, 8, 7, 2, 0, 6, 5, 10, 99, -20, 34};
    mergeSort(nums, 0, nums.size()-1);
    for(auto num: nums)
        cout<<num<<' ';
    cout<<endl;
    return 0;
}